### ISSUE TITLE
Street names rendering improvements

### ISSUE THREAD

**Issue by pbmorton** [mapsforge/mapsforge#644]
Using mapsforge 0.5.1 sample App.  If you change one rule in rendertheme-v4.xml to the following (the only change in this rule is zoom-min=15 changed to zoom-min=13):

    <rule e="way" k="highway" v="secondary_link|primary_link|trunk_link|motorway_link|secondary|primary" zoom-min="13">
    <pathText k="name" display="always" priority="-6" font-style="bold" font-size="12" fill="#000000" stroke="#FFFFFF" stroke-width="2.0"/>
    </rule>

You should see highway names on the map at zoom 13 and 14.  But they do not appear until zoom=15, even though the highways are visible at zooms 13 and 14.

**Comment by devemux86** [comment]
Line labels are a special case.
Besides all other placement rules, here internally we have to make sure that each label "fits" in the line segment it's supposed to appear. If it is too long it's omitted.

**Comment by pbmorton** [comment]
Thanks.  From your response it seems this is expected behaviour.  I think you should consider changing this behaviour in future.  At zoom=13 (for example) a map may only show major highways and it would be useful to display the names of the highways.  Forcing a user to zoom in to get the names to appear, then zoom out again to use the map, is not user friendly in my opinion.  It is inconsistent too as other features like rivers display names at zoom=13. What do you think about changing this in future?

**Comment by devemux86** [comment]
I agree that this is not the best behavior and we should review it.

**Comment by devemux86** [comment]
BTW have you played with [render theme](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md) directives, like ```priority``` and ```display```?

**Comment by pbmorton** [comment]
I did play around with the display and priority directives but could not make the labels visible.

**Comment by eartrumpet** [comment]
The only way to make highway names visible at lower zoom levels (sometimes, if the highway has short segments and a long name, this means ZL 16-18!) is to use caption instead of pathText - with other annoyances involved.

This is one of the last major disadvantages of mapsforge which regularly creates support questions.

Can this be fixed with the label layer? OSMAnd for example has pathTexts winding along the highway, which would be desireable too.

**Comment by devemux86** [comment]
>Can this be fixed with the label layer?

Have you tried it?

**Comment by eartrumpet** [comment]
At the moment it makes no difference, and it seems to me that pathText is still rendered on the map layer.

Right now pathText is rendered on every possible segment, if you zoom in far enough, which isn't desireable as well (especially with possible solution like truncating or just writing the name longer than the segment) .
Also only once per tile if it's a long straight segment which passes several tiles (for example on aerial ways). 

Could moving to label layer it might be possible to limit it to once per name value? Then it might make sense to just write the whole name, even if it's longer than a segment.

**Comment by Klaus-Tockloth** [comment]
emux already referenced it ... so only to keep the context ... here's the relevant screenshot (which shows a lot of missing street labels).

![bildschirmfoto 2016-02-07 um 20 18 09](https://cloud.githubusercontent.com/assets/5493128/12879575/706c15c8-ce35-11e5-8d6a-11aaa27e4422.jpg)

It seems that the rules for supressing the street labels are to strict. IMHO it's important to increase the number of labels shown, even if they are not rendered perfectly (or have some glitches).

Klaus

**Comment by ludwigb** [comment]
I changed the algorithm a bit:
 - moving street names into the middle of a segment 
 - reduce margins where I think they were overly cautions or even double calculated
 - reduce margin value
Overall we get more street names with a more pleasing layout.

**Comment by devemux86** [comment]
+1

Are we certain the labels are in segments center, they seem a little off.

Also the `skipPixels` - specially now with arbitrary value - what's exactly its purpose?

**Comment by ludwigb** [comment]
skipPixels is used to avoid repeating the street name all the time if there
is space. Previously it was dependent on the name length, which to me does
not make much sense.

actually, just double checking: the calculations might be total rubbish,
will have to check tomorrow or so again. But at least we do not have them
stuck at the beginning of the road anymore.

Thanks for checking....

**Comment by devemux86** [comment]
Another thought is if we need to make `skipPixels` dependent on display model scale factor now.

**Comment by ludwigb** [comment]
possibly, I was wondering what unit that actually is.

**Comment by lincomatic** [comment]
This new code shows more streets, which is great, but at the same time, there are more duplicate names on the same street.

Old code:
![image](https://cloud.githubusercontent.com/assets/1422186/14757227/fd179e70-08a5-11e6-89cb-aed72de61741.png)

New code:
![image](https://cloud.githubusercontent.com/assets/1422186/14757240/2acfadb2-08a6-11e6-85b0-b4b23fe50a3a.png)

**Comment by ludwigb** [comment]
yes, that is what the code does: it relaxes the space restrictions for
street names a bit, so sections that were not considered long enough now
sometimes display a street name

**Comment by lincomatic** [comment]
I fixed a bug in https://github.com/mapsforge/mapsforge/commit/c8f705f6f635c764cec684cb8cface4f8dec691c . The street names were not centered. Compare the screen shot below with the one above:
![image](https://cloud.githubusercontent.com/assets/1422186/14805058/7fd216b4-0b1d-11e6-9e0a-3abdae4ecfcb.png)

In WayDecorator.java, the line:

`            LineSegment actuallyUsedSegment = drawableSegment.subSegment(offset, wayNameWidth - WAYNAME_SAFETY_MARGIN * 2);
`

should be:

`            LineSegment actuallyUsedSegment = drawableSegment.subSegment(offset+WAYNAME_SAFETY_MARGIN, wayNameWidth - WAYNAME_SAFETY_MARGIN * 2);
`

**Comment by lincomatic** [comment]
Sorry, my previous comment was referencing lines in stale code. I submitted a PR with the latest dev code https://github.com/mapsforge/mapsforge/pull/829

**Comment by devemux86** [comment]
Closing this part of improvements. Additional can be monitored in new issue.
